### PR TITLE
Make sure `AssemblyDefinitions` are disposed after use.

### PR DESCRIPTION
**Pull request by TimonPost**
NUnit does not correctly release the file lock for some of the assemblies. This PR makes sure the `Dispose` functionality is called thereby releasing all the underlying file resources.

### Mono Cecil (Fixed in this PR)
 `Mono.Cecil.AsseblyDefinition` is not disposed of, this locks the assembly until the caller application closes.

- https://github.com/nunit/nunit-console/blob/master/src/NUnitEngine/nunit.engine.core/Services/DriverService.cs#L76
- https://github.com/nunit/nunit-console/blob/master/src/NUnitEngine/nunit.engine/Services/RuntimeFrameworkService.cs#L233
- https://github.com/nunit/nunit-console/blob/master/src/NUnitEngine/nunit.engine.core/Drivers/NUnitNetStandardDriver.cs#L84
- https://github.com/nunit/nunit-console/blob/master/src/NUnitEngine/nunit.engine.core/Extensibility/ExtensionAssembly.cs#L37

### CustomAssemblyLoadContext (Not fixed in this PR)

Currently, an assembly is loaded by this [code](https://github.com/nunit/nunit-console/blob/master/src/NUnitEngine/nunit.engine.core/Drivers/NUnitNetCore31Driver.cs#L84). 

```csharp
var assemblyLoadContext = new CustomAssemblyLoadContext(assemblyPath);
_testAssembly = assemblyLoadContext.LoadFromAssemblyPath(assemblyPath);
```

Next, the following private fields are initialized:

```csharp
Assembly _testAssembly;
Assembly _frameworkAssembly;
object _frameworkController;
Type _frameworkControllerType;
```

Those have to be released in order for `CustomAssemblyLoadContext.Unload()` to be able to release its resources. 

_microsoft [article](https://docs.microsoft.com/en-us/dotnet/standard/assembly/unloadability):_
image

### Related Issues
Fixes: #894 

### Work Left for other PR
- Make `CustomAssemblyLoadContext.Unload()` working see [this article](https://docs.microsoft.com/en-us/dotnet/standard/assembly/unloadability). Lets discuss this issue in the related issue.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/nunit/nunit-console?pullRequest=895) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/nunit/nunit-console?pullRequest=895) before we can accept your contribution.<br/><br/>:x: TimonPost [sign now](https://cla.dotnetfoundation.org/nunit/nunit-console?pullRequest=895)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/nunit/nunit-console?pullRequest=895) it.</sub>

**Comment by CharliePoole** [comment]
@TimonPost As I mentioned on the issue, the entire reason we use `AssemblyDefinition`, which I extracted and modified from `Mono.Cecil`, is to __avoid__ loading the assembly. So while your fix may work for the immediate problem, it bothers me to do it without understanding why it's really necessary. One thing that would help would be to have a test that fails before and passes after your fix. If my `AssemblyDefinition` is holding the assembly loaded, I would prefer to fix that directly in the metadata assembly if possible.

Note that I'm only talking about the AssemblyDefinition part of this. I didn't work on the .NET Core driver, so I'll leave it to @ChrisMaddock to comment on that.

**Comment by TimonPost** [comment]
The `AssemblyDeffinitions` in the following files are only used locally in method scope, therefore I think it's perfect to dispose of them directly after use:

NUnitNetStandardDriver.cs
DriverService.cs 
RuntimeFrameworkService.cs


The `AssemblyDeffinition`  in the `ExtensionAssembly.cs` is a bit different since it is stored as a local variable, and exposes public member to request access to various parts of this `AssemblyDefinition`. I am not sure if it is the correct way to implement `Disposable` for this class. Any ideas?

**Comment by CharliePoole** [comment]
I've been wanting to get rid of ExtensionAssembly, but that's probably a bigger job than is called for by this issue. I agree with the approach you have taken by making it disposable, but we need to make sure that the code using it actually calls Dispose. I'll do a more in-depth code review tomorrow ... well today, but in the morning. :smile:

**Comment by TimonPost** [comment]
Okay, I think it might be a nice idea to separate the `AssemblyDefinition` issue and `CustomAssemblyLoadContext` into two PR's since resolving the `CustomAssemblyLoadContext` issue sounds a bit more difficult.

**Comment by CharliePoole** [comment]
I like the use of StopService - that's what it's for. :smile: 

Since you plan to modify this PR, post a comment on it when it's ready for review and we'll see it.

**Comment by TimonPost** [comment]
I pushed the changes which introduce `StopService`, also every using statement that is implemented. Unfortunately, those using statements cause code to be intended which makes it look like more changes than it actually is. I think it is ready for review. Removed the WIP status.

**Comment by CharliePoole** [comment]
Going back to bed now. :wink: I'll go through the code tomorrow.

**Review by CharliePoole — APPROVED** [review]
This looks good to me. You'll need one more committer to approve it before we merge. Thanks for your work!

**Comment by ChrisMaddock** [comment]
This looks good - thanks for digging into this Timon! Will discuss the remaining problem on the issue.

@CharliePoole - this sort of issue is an interesting argument towards including the Mono.Cecil files directly in the engine, which didn't really cross my mind when we were considering it before. Maybe something to consider for the future. 🙂

**Comment by CharliePoole** [comment]
@ChrisMaddock Not sure why you say that about including directly. The same issue exists no matter how the assembly is used.

**Pull request merged**
